tests/loops/test_evaluation_loop.py

```diff
@@ -42,6 +44,37 @@ def test_on_evaluation_epoch_end(eval_epoch_end_mock, tmpdir):
     assert eval_epoch_end_mock.call_count == 4
 
 
+def test_set_epoch_called_eval_predict(tmpdir):
+    """Tests that set_epoch (if the sampler has one) is called on the DataLoader during evaluation and
+    prediction."""
+
+    def _get_dataloader():
+        dataset = RandomDataset(32, 64)
+        sampler = RandomSampler(dataset)
+        sampler.set_epoch = Mock()
+        sampler.set_epoch.reset_mock()
+        return DataLoader(dataset, sampler=sampler)
+
+    model = BoringModel()
+
+    train_dataloader = _get_dataloader()
+    val_dataloader = _get_dataloader()
+
+    trainer = Trainer(
+        default_root_dir=tmpdir, limit_train_batches=2, limit_val_batches=2, max_epochs=2, enable_model_summary=False
+    )
+
+    trainer.fit(model, train_dataloaders=train_dataloader, val_dataloaders=val_dataloader)
```

Comment:
We need to pass the data loaders (for training and validation) separately else the model's data loader is considered for evaluation. 

https://github.com/PyTorchLightning/pytorch-lightning/blob/ae4bd0354548d1234dbbfb2acce95d6323ec6b91/pytorch_lightning/trainer/connectors/data_connector.py#L187-L189

This fixes the error, and `set_epoch` now works as expected. :)